Mathlib/NumberTheory/LSeries/AbstractFuncEq.lean

```diff
@@ -0,0 +1,450 @@
+
+/-!
+## Auxiliary results I: lemmas on asymptotics
+-/
+
+/-- As `x → 0`, we have `F x = x ^ (-P.k) • constant` up to a rapidly decaying error. -/
+lemma hf_zero (P : WeakFEPair E) (r : ℝ) :
+    (fun x ↦ P.f x - (P.ε * ↑(x ^ (-P.k))) • P.g₀) =O[𝓝[>] 0] (· ^ (-r)) := by
+  have := (P.hg_top (-(r - P.k))).comp_tendsto tendsto_inv_zero_atTop
+  simp_rw [IsBigO, IsBigOWith, eventually_nhdsWithin_iff] at this ⊢
+  obtain ⟨C, hC⟩ := this
+  use ‖P.ε‖ * C
+  filter_upwards [hC] with x hC' (hx : 0 < x)
+  have h_nv2 : ↑(x ^ P.k) ≠ (0 : ℂ) := ofReal_ne_zero.mpr (rpow_pos_of_pos hx _).ne'
+  have h_nv : P.ε⁻¹ * ↑(x ^ P.k) ≠ 0 := mul_ne_zero P.symm.hε h_nv2
+  specialize hC' hx
+  simp_rw [Function.comp_apply, ← one_div, neg_neg, P.h_feq' _ hx] at hC'
+  rw [← ((mul_inv_cancel h_nv).symm ▸ one_smul ℂ P.g₀ :), mul_smul _ _ P.g₀, ← smul_sub, norm_smul,
+    ← le_div_iff' (lt_of_le_of_ne (norm_nonneg _) (norm_ne_zero_iff.mpr h_nv).symm)] at hC'
+  convert hC' using 1
+  · congr 3
+    rw [rpow_neg hx.le]
+    field_simp
+  · simp_rw [norm_mul, norm_real, rpow_neg hx.le, one_div, inv_rpow hx.le, norm_inv,
+      norm_of_nonneg (rpow_pos_of_pos hx _).le, rpow_sub hx]
+    field_simp
+    ring
+
+/-- Power asymptotic for `f - f₀` as `x → 0`. -/
+lemma hf_zero' (P : WeakFEPair E) :
+    (fun x : ℝ ↦ P.f x - P.f₀) =O[𝓝[>] 0] (· ^ (-P.k)) := by
+  simp_rw [← fun x ↦ sub_add_sub_cancel (P.f x) ((P.ε * ↑(x ^ (-P.k))) • P.g₀) P.f₀]
+  refine (P.hf_zero _).add (IsBigO.sub ?_ ?_)
+  · rw [← isBigO_norm_norm]
+    simp_rw [mul_smul, norm_smul, mul_comm _ ‖P.g₀‖, ← mul_assoc, norm_real]
+    apply (isBigO_refl _ _).const_mul_left
+  · refine IsBigO.of_bound ‖P.f₀‖ (eventually_nhdsWithin_iff.mpr ?_)
+    filter_upwards [eventually_le_nhds zero_lt_one] with x hx' (hx : 0 < x)
+    apply le_mul_of_one_le_right (norm_nonneg _)
+    rw [norm_of_nonneg (rpow_pos_of_pos hx _).le, rpow_neg hx.le]
+    exact one_le_inv (rpow_pos_of_pos hx _) (rpow_le_one hx.le hx' P.hk.le)
+
+end WeakFEPair
+
+namespace StrongFEPair
+
+variable (P : StrongFEPair E)
+
+/-- As `x → 0`, `F x` decays faster than any power of `x`. -/
+lemma hf_top' (r : ℝ) :
+    P.f =O[atTop] (· ^ (-r)) := by
+  simpa only [P.hf₀, sub_zero] using P.hf_top r
+
+/-- As `x → 0`, `F x` decays faster than any power of `x`. -/
+lemma hf_zero' (r : ℝ) :
+    P.f =O[𝓝[>] 0] (· ^ (-r)) := by
+  have := P.hg₀ ▸ P.hf_zero r
+  simpa only [smul_zero, sub_zero]
+
+/-!
+## Main theorems on strong FE-pairs
+-/
+
+/-- The completed L-function. -/
+def Λ : ℂ → E := mellin P.f
+
+/-- The Mellin transform of `f` is well-defined and equal to `P.Λ s`, for all `s`. -/
+lemma hasMellin (s : ℂ) :
+    HasMellin P.f s (P.Λ s):=
+  let ⟨t, ht⟩ := exists_gt s.re
+  let ⟨u, hu⟩ := exists_lt s.re
+  ⟨mellinConvergent_of_isBigO_rpow P.hf_int (P.hf_top' t) ht (P.hf_zero' u) hu, rfl⟩
+
+/-- If `(f, g)` are a strong FE pair, then the Mellin transform of `f` is entire. -/
+lemma differentiable_Λ : Differentiable ℂ P.Λ := fun s ↦
```

Comment:
```suggestion
lemma Differentiable_Λ : Differentiable ℂ P.Λ := fun s ↦
```

Reply:
I don't think that's the convention, is it? `Differentiable` is `Type`-valued, but `differentiable_Λ` is term-valued (it constructs a term of type `Differentiable xxx`) so it is lowercased.

Reply:
Yep, good point. Sorry for the noise.